fix: fix focus behavior of Layer in presence of focusable contents

crashuniverse opened the pull request:
Fixes #3112

Signed-off-by: Priya Ranjan Singh <mail@priyaranjan.net>

<!--- Provide a general summary of the PR in the Title above -->

#### What does this PR do?
This PR introduces conditions to check for already focused elements in contents of Layer and avoids stealing focus if there is any. In absence of any focused elements in contents of Layer, it continues to assign focus to the hidden anchor element.

#### Where should the reviewer start?
`src/js/components/Layer/LayerContainer.js` has changes in `componentDidMount` lifecycle.

#### What testing has been done on this PR?
In the `Form` story that renders `<FormLayer />`, placing an `autoFocus` to `<TextInput />` component. The component finally looks to be `<TextInput autoFocus />`. When layer container is visible, `TextInput` has focus. In absence of `autoFocus` prop, the hidden anchor continues to have the focus.

#### How should this be manually tested?
On this branch, if we add `autoFocus` prop in story https://storybook.grommet.io/?path=/story/layer--form, to the input element, we would see that it received the focus when Layer opens.

#### Any background context you want to provide?
This problem is observed initially with several open source modals. It is referred to as stealing autofocus.

#### What are the relevant issues?
#3112

#### Screenshots (if appropriate)
No

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
No

#### Is this change backwards compatible or is it a breaking change?
Yes, the change is backwards compatible.

oorestisime commented:
don't have time to properly review this before next week but just a quick note. we are missing the documentation of the new prop. you need to add this in the doc.js file and don't forget about the ts typings as well :)

crashuniverse commented:
Perfect. I wasn't familiar with these two related things. I'll get on it. Thanks.

crashuniverse commented:
Updated `doc.js`, generated supporting documentation & updated typescript definitions.

ericsoderberghp commented:
Rather than having to add a property, I think we could detect whether something inside the Layer already has the focus and only focus on the anchor if that isn't the case. Perhaps:
```
  componentDidMount() {
    const { position } = this.props;
    if (position !== 'hidden') {
      this.makeLayerVisible();
      // Once layer is open we make sure it has focus so that you
      // can start tabbing inside the layer. If the caller put focus
      // on an element already, we honor that. Otherwise, we put
      // the focus in the hidden anchor.
      let element = document.activeElement;
      while (element) {
        if (element === this.containerRef.current) {
          // already have focus inside the container
          break;
        }
        element = element.parentElement;
      }
      if (!element && this.anchorRef.current) {
        this.anchorRef.current.focus();
      }
    }
  }
```

crashuniverse commented:
I like your approach. I'm going to give this try this weekend. Thank you so much for sharing your thoughts.

crashuniverse commented:
I updated the approach, thank you for feeding me the solution. I kept the tests I wrote before as even when I changed the approach, I found them relevant. I'd request you to to check my test where I make it look for an element 'A' as activeElement, something that I think is okay but would like to have it double checked. I'd return the favor by fixing another a11y bug with the new acquired knowledge of `activeElement` and `element.parent`. Thank you again.

ericsoderberghp commented on src/js/components/Layer/__tests__/Layer-test.js:
```diff
@@ -212,4 +212,34 @@ describe('Layer', () => {
+    /* eslint-disable jsx-a11y/no-autofocus */
+    const layerNode = getByTestId(document, 'focus-input-test');
+    expect(layerNode).toMatchSnapshot();
+    expect(document.activeElement.nodeName).toBe('INPUT');
```
It would be tighter to give the input a `data-testid` and validate against that.

crashuniverse replied:
Much better. At time of writing I realized there could be other possible ways of identifying this element better, so I made it more specific with a `datat-testid`.

ericsoderberghp commented on src/js/components/Layer/__tests__/Layer-test.js:
```diff
@@ -212,4 +212,34 @@ describe('Layer', () => {
+
+    const layerNode = getByTestId(document, 'focus-layer-test');
+    expect(layerNode).toMatchSnapshot();
+    expect(document.activeElement.nodeName).toBe('A');
```
That's probably OK.

It might be slightly more robust to add an `<input autofocus />` after the `Layer`. Just a thought.

crashuniverse replied:
Yes, that would make it more hostile environment, so I added it.

crashuniverse replied:
Notify @ericsoderberghp

ericsoderberghp commented:
Nicely done. Thanks for your contribution!

ShimiSun commented:
I've released grommet 2.7.3 to include this fix, THANK YOU for the contribution @crashuniverse!

crashuniverse commented:
Thank you @ShimiSun :)
